lib/segment/src/vector_storage/mmap_type.rs

```diff
@@ -0,0 +1,504 @@
+///
+/// # Safety
+///
+/// - unsafe because we create a second (unbounded) mutable reference
+/// - malformed data in the mmap may break the transmuted type `T` resulting in undefined behavior
+///
+/// # Panics
+///
+/// - panics when the size of the mmap doesn't match size `T`
+/// - panics when the mmap data is not correctly aligned for type `T`
+unsafe fn mmap_to_type_unbounded<'unbnd, T>(mmap: &mut MmapMut) -> &'unbnd mut T
+where
+    T: Sized,
+{
+    // Obtain unbounded bytes slice into mmap
+    let bytes: &'unbnd mut [u8] = {
+        let slice = mmap.deref_mut();
+        slice::from_raw_parts_mut(slice.as_mut_ptr(), slice.len())
+    };
+
+    // Assert alignment and size
+    assert_alignment::<_, T>(bytes);
+    debug_assert_eq!(mmap.len(), bytes.len());
+    assert_eq!(bytes.len(), mem::size_of::<T>());
+
+    let ptr = bytes.as_ptr() as *mut T;
+    unsafe { &mut *ptr }
+}
+
+/// Get a second mutable reference for a slice of type `T` from the given mmap
+///
+/// A (non-zero) header size in bytes may be provided to omit from the BitSlice data.
+///
+/// # Warning
+///
+/// The returned reference is unbounded. The user must ensure it never outlives the `mmap` type.
+///
+/// # Safety
+///
+/// - unsafe because we create a second (unbounded) mutable reference
+/// - malformed data in the mmap may break the transmuted slice for type `T` resulting in undefined
+///   behavior
+///
+/// # Panics
+///
+/// - panics when the size of the mmap isn't a multiple of size `T`
+/// - panics when the mmap data is not correctly aligned for type `T`
+/// - panics when the header size isn't a multiple of size `T`
+unsafe fn mmap_to_slice_unbounded<'unbnd, T>(
+    mmap: &mut MmapMut,
+    header_size: usize,
+) -> &'unbnd mut [T]
+where
+    T: Sized,
+{
+    // Obtain unbounded bytes slice into mmap
+    let bytes: &'unbnd mut [u8] = {
+        let slice = mmap.deref_mut();
+        &mut slice::from_raw_parts_mut(slice.as_mut_ptr(), slice.len())[header_size..]
+    };
+
+    // Assert alignment and size
+    assert_alignment::<_, T>(bytes);
+    debug_assert_eq!(mmap.len(), bytes.len() + header_size);
+    debug_assert_eq!(
+        header_size % mem::size_of::<T>(),
+        0,
+        "header not multiple of size T",
+    );
+    assert_eq!(
+        bytes.len() % mem::size_of::<T>(),
+        0,
+        "bytes not multiple of size T",
+    );
+
+    // Transmute slice types
+    slice::from_raw_parts_mut(bytes.as_ptr() as *mut T, bytes.len() / mem::size_of::<T>())
+}
+
+/// Assert slice `&[S]` is correctly aligned for type `T`.
+///
+/// # Panics
+///
+/// Panics when alignment is wrong.
+fn assert_alignment<S, T>(bytes: &[S]) {
+    // We do not enforce alignment on Windows at this time, because the Windows API doesn't
+    // guarantee memory mapped segments to be aligned.
+    //
+    // We should look into this again to confirm this is sound.
```

Comment:
Producing a reference to unaligned data, even if not reading or writing through that reference, is UB: https://doc.rust-lang.org/reference/behavior-considered-undefined.html

Reply:
This is a hard problem. We discovered that mmap'ed files are not aligned on Windows.

We discussed this, and didn't find an easy solution to mitigate this. The very same is already in other places. We decided to match that for now, which means, disabling this check on Windows, until we find a better solution.

Do you have a good suggestion for this? We might even end up disabling mmap support on Windows+ARM builds.

Reply:
Windows (at least newer versions) has API's to create aligned mmaps and in the cases I can see, it by default aligns to 64kB boundaries on my system. Which cases did you find where the map is not aligned? Is it only on ARM?

Reply:
Thanks for your reply.

On Windows it fails in unit tests with a very basic alignment check.

### What happens

This is already merged in the `dev` branch, so we can use that as base.

If we enable the alignment check on Windows here:

https://github.com/qdrant/qdrant/blob/83f8b50acadd25d47a9bdced1e762b924a7b272d/lib/segment/src/common/mmap_type.rs#L463-L472

This simple check fails (amongst others):

https://github.com/qdrant/qdrant/blob/83f8b50acadd25d47a9bdced1e762b924a7b272d/lib/segment/src/common/mmap_type.rs#L498

Here's the error on CI when it was enabled: https://github.com/qdrant/qdrant/actions/runs/4923938726/jobs/8796365496#step:6:1213

If you enable the assertion on Windows as noted above, you can probably easily run the test with: `cargo test --profile dev --package segment test_open-zero`

All it does is:
- Create file on disk (with various sizes)
- Open the file (read/write)
- Open mmap with [`MmapMut::map_mut`](https://docs.rs/memmap2/latest/memmap2/struct.MmapMut.html#method.map_mut)
- Then our alignment assertion complains the bytes in memory are off-by-one (`bytes.as_ptr().align_offset(mem::align_of::<T>()) == 7`)

Either the file isn't mapped in a 64k grid, or the check is somehow faulty. I don't have good knowledge of Windows APIs.

---

> Is it only on ARM?

This is on CI. Which is x86_64.

Reply:
Thanks for the detailed reply. I'm intrigued, so I'll look into it a bit and let you know what I find. I have a feeling the actual windows syscall returns an aligned pointer and some pointer arithmetic in memmap2 makes it unaligned afterwards, but for the moment that feeling is completely unsubstantiated.

Reply:
That would be fantastic! Let me know if I can be of any help. I'd be happy to prepare a more concise example in code for you.

Reply:
That makes a lot of sense. Thanks a lot for taking the time to investigate this and for figuring this out. That's super useful! A bit of a bummer that the crate doesn't return this edge case as an error, that would have made debugging a lot easier.

I'll fix this tomorrow, adding a check to mitigate this very situation.

It makes sense it fails, because in the test we actually create a file with `.set_len(0)`.

Thanks once again!